Src/WindowsServer/WindowsServer.Shared/FirstChanceExceptionStatisticsTelemetryModule.cs

```diff
@@ -0,0 +1,188 @@
+        private ConcurrentBag<string> methodValues = new ConcurrentBag<string>();
+        private ConcurrentBag<string> typeValues = new ConcurrentBag<string>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FirstChanceExceptionStatisticsTelemetryModule" /> class.
+        /// </summary>
+        public FirstChanceExceptionStatisticsTelemetryModule() : this(
+            action => AppDomain.CurrentDomain.FirstChanceException += action,
+            action => AppDomain.CurrentDomain.FirstChanceException -= action)
+        {
+        }
+
+        internal FirstChanceExceptionStatisticsTelemetryModule(
+            Action<EventHandler<FirstChanceExceptionEventArgs>> registerAction,
+            Action<EventHandler<FirstChanceExceptionEventArgs>> unregisterAction)
+        {
+            this.registerAction = registerAction;
+            this.unregisterAction = unregisterAction;
+        }
+
+        /// <summary>
+        /// Initializes the telemetry module.
+        /// </summary>
+        /// <param name="configuration">Telemetry Configuration used for creating TelemetryClient for sending exception statistics to Application Insights.</param>
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+            // Core SDK creates 1 instance of a module but calls Initialize multiple times
+            if (!this.isInitialized)
+            {
+                lock (this.lockObject)
+                {
+                    if (!this.isInitialized)
+                    {
+                        this.isInitialized = true;
+
+                        this.telemetryClient = new TelemetryClient(configuration);
+                        this.telemetryClient.Context.GetInternalContext().SdkVersion = SdkVersionUtils.GetSdkVersion("exstat:");
+
+                        this.metricManager = new MetricManager(this.telemetryClient);
+
+                        this.registerAction(this.CalculateStatistics);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Disposing TaskSchedulerOnUnobservedTaskException instance.
+        /// </summary>
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// IDisposable implementation.
+        /// </summary>
+        /// <param name="disposing">The method has been called directly or indirectly by a user's code.</param>
+        private void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                this.unregisterAction(this.CalculateStatistics);
+                this.metricManager.Dispose();
+            }
+        }
+
+        private void CalculateStatistics(object sender, FirstChanceExceptionEventArgs firstChanceExceptionArgs)
+        {
+            if (executionSyncObject == LOCKED)
```

Comment:
Is it just a protection from StackOverflow, so it's guaranteed that we already passed this method fully once (up until exception in our code) and won't just pass again?